modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,492 @@
+//            ]
+//        }
+//    }
+//
+// The following path accesses the value "3":
+//
+//    foo.bars.0.baz
+//
+// The path is accompanied by a new object. Thus, to increment the value "3"
+// in the above object, we would use the following journalUpdate:
+//
+//    {
+//        "p": "foo.bars.0.baz",
+//        "v": 4
+//    }
+//
+// All permutations of the journalUpdate object are legal. However, malformed updates
+// are ignored during application. A journalUpdate is considered malformed in three
+// circumstances:
+//
+// - Its Path references an element that does not exist at application time.
+//   This includes out-of-bounds array indices.
+// - Its Path contains invalid characters (e.g. "). See the JSON spec.
+// - Value contains invalid JSON or is empty.
+//
+// Other special cases are handled as follows:
+//
+// - If Path is "", the entire object is replaced.
+// - If an object contains duplicate keys, the first key encountered is used.
+//
+// Finally, to enable efficient array updates, the length of the array (at
+// application time) may be used as a special array index.  When this index is
+// the last accessor in Path, Value will be appended to the end of the array.
+// If the index is not the last accessor, the journalUpdate is considered malformed
+// (and thus is ignored).
+type journalUpdate struct {
+	// Path is an arbitrarily-nested JSON element, such as foo.bars.1.baz
+	Path string `json:"p"`
+	// Value contains the new value of Path.
+	// TODO: remove pointer once Go 1.8 is released.
+	Value *json.RawMessage `json:"v"`
+}
+
+// apply applies u to obj, returning the new JSON, which may share underlying
+// memory with obj or u.Value. If u is malformed, obj is returned unaltered.
+// See the journalUpdate docstring for an explanation of malformed journalUpdates. If obj is
+// not valid JSON, the result is undefined.
+func (u journalUpdate) apply(obj json.RawMessage) json.RawMessage {
+	if len(*u.Value) == 0 {
+		// u is malformed
+		return obj
+	}
+	return rewritePath(obj, u.Path, *u.Value)
+}
+
+// newJournalUpdate constructs an update using the provided path and val. If val
+// cannot be marshaled, newJournalUpdate panics. If val implements the json.Marshaler
+// interface, it is called directly. Note that this bypasses validation of the
+// produced JSON, which may result in a malformed journalUpdate.
+func newJournalUpdate(path string, val interface{}) journalUpdate {
+	var data []byte
+	var err error
+	if m, ok := val.(json.Marshaler); ok {
+		// bypass validation
+		data, err = m.MarshalJSON()
+	} else {
+		data, err = json.Marshal(val)
+	}
+	if err != nil {
+		panic(err)
+	}
+	rm := json.RawMessage(data)
+	return journalUpdate{
+		Path:  path,
+		Value: &rm,
+	}
+}
+
+// rewritePath replaces the value at path in json with val. The returned slice
+// may share underlying memory with json. If path is malformed, the original
+// json is returned.
+func rewritePath(json []byte, path string, val []byte) []byte {
```

Comment:
definitely need to be returning an error or dropping a build.Critical for malformed stuff (probably both). Silently dropping updates is not okay.